_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+
+```
+├ 0     msg.value > 0
+│       Revert
+│
+└ 1     msg.value == 0
+  │
+  ├ 0   x + y < x
+  │     Revert("overflow!")
+  │
+  └ 1   x + y >= x
+        Return: x + y
+```
+
+Each leaf on the tree represents a possible execution path, and has some logical conditions
+attached:
+
+- `0`:  `msg.value > 0`: revert (`add` is not payable)
+- `10`: `msg.value == 0 && x + y < x`: revert (overflow)
+- `11`: `msg.value == 0 && x + y >= x`: return `x + y`
+
+We can therefore use symbolic execution to exhaustively explore all possible execution paths for a
+given program. This means we can write tests that assert that properties hold for *all possible
+inputs* to a given function.
+
+## Using `ds-test`
+
+[`ds-test`](https://github.com/dapphub/ds-test/blob/master/src/test.sol) is a solidity unit testing
+libary with tight integration into `hevm`. `hevm` will execute as a test any method that meets the
+following criteria:
+
+1. The method is on a contract that inherits from `DSTest`
+1. The method is `public` or `external`
+1. The method name starts with `test` or `prove`
+
+If a public method named `setUp()` is present it will be executed before each test is run.
+
+Three types of test are recognised by `hevm`:
+
+| type           | prefix  | arguments | semantics                                                   |
+|----------------|---------|-----------|-------------------------------------------------------------|
+| concrete       | `test`  | no        | single execution with concrete values                       |
+| property based | `test`  | yes       | multiple executions with randomly generated concrete values |
+| symbolic       | `prove` | optional  | exhaustive exploration of all possible exeuction paths      |
+
+To illustrate the differences between the test types, consider the following example:
+
+```solidity
+pragma solidity ^0.6.12;
+import {DSTest} from "ds-test/test.sol";
+
+contract Test is DSTest {
+    function test_associativity() public {
+        assertEq(
+            uint((1 + 2) + 3),
+            uint(1 + (2 + 3))
+        );
+    }
+
+    function test_associativity_fuzz(uint x, uint y, uint z) public {
+        assertEq(
+            (x + y) + z,
+            x + (y + z)
+        );
+    }
+
+    function prove_associativity(uint x, uint y, uint z) public {
+        assertEq(
+          (x + y) + z,
+          x + (y + z)
+        );
+    }
+}
+```
+
+- `test_associativity` will run a single test case with the concrete values of `1`, `2`, and `3`.
+- `test_associativity_fuzz` will be executed many times (100 by default), with randomly generated values for `x`, `y`, and `z`.
+- `prove_associativity` will be symbolically executed, with `x`, `y`, and `z` represented as symbolic variables.
+
+Finally, it is possible to manipulate the execution environment (e.g. timestamp) from within
+`ds-test` using by using hevm "[cheat codes](https://github.com/dapphub/dapptools/tree/master/src/hevm#cheat-codes)"
+
```

Comment:
We could expand this a little to its own section, maybe along the lines of:
### Setting up the environment
- We can spin up the contracts we want, or 
- use rpc to get real state, and
- tweak the environment with cheat codes